trace_decoder/src/decoding.rs

```diff
@@ -21,22 +21,122 @@ use mpt_trie::{
 use thiserror::Error;
 
 use crate::{
-    processed_block_trace::{NodesUsedByTxn, ProcessedBlockTrace, StateTrieWrites, TxnMetaState},
+    processed_block_trace::{
+        NodesUsedByTxn, ProcessedBlockTrace, ProcessedTxnInfo, StateTrieWrites, TxnMetaState,
+    },
     types::{
         HashedAccountAddr, HashedNodeAddr, HashedStorageAddr, HashedStorageAddrNibbles,
         OtherBlockData, TriePathIter, TrieRootHash, TxnIdx, TxnProofGenIR,
         EMPTY_ACCOUNT_BYTES_RLPED, ZERO_STORAGE_SLOT_VAL_RLPED,
     },
-    utils::{hash, update_val_if_some},
+    utils::{hash, update_val_if_some, write_optional},
 };
 
 /// Stores the result of parsing tries. Returns a [TraceParsingError] upon
 /// failure.
-pub type TraceParsingResult<T> = Result<T, TraceParsingError>;
+pub type TraceParsingResult<T> = Result<T, Box<TraceParsingError>>;
+
+/// Represents errors that can occur during the processing of a block trace.
+///
+/// This struct is intended to encapsulate various kinds of errors that might
+/// arise when parsing, validating, or otherwise processing the trace data of
+/// blockchain blocks. It could include issues like malformed trace data,
+/// inconsistencies found during processing, or any other condition that
+/// prevents successful completion of the trace processing task.
+#[derive(Debug)]
+pub struct TraceParsingError {
+    block_num: Option<U256>,
+    block_chain_id: Option<U256>,
+    txn_idx: Option<usize>,
+    addr: Option<Address>,
+    h_addr: Option<H256>,
+    slot: Option<U512>,
+    slot_value: Option<U512>,
+    reason: TraceParsingErrorReason, // The original error type
+}
+
+impl std::fmt::Display for TraceParsingError {
+    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
```

Comment:
We can probably do this in a single `write` in a nicer way.

Reply:
I have an idea how to do it. Let me know if you want to discuss it

Reply:
let's discuss it